### PR TITLE
Impl source instead of cause for error::Error

### PR DESCRIPTION
**Pull request by Gbillou**
Switch from `cause` to `source` method for `std::error::Error` impls.

### CODE REVIEW

**Review by jonhoo — CHANGES_REQUESTED** [review]
Oh, yes, this is a good change to make! However, I think all of these changes should continue to return `Some(..)` rather than calling `.source()` internally, as otherwise you'd "miss" the middle error type in the source chain.

**Comment by Gbillou** [comment]
I agree but `Display for Error` (and their deprecated `description()` counter part) do forward to their inner error.

I believe either we keep returning `Some(se)` for `source()` but also update impl Display so tokio-tower errors do return their own description (using pipeline/server as an example):
```
impl<T, S> fmt::Display for Error<T, S>
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Error::BrokenTransportRecv(ref se) => f.write_str("The underlying transport failed to produce a request"),
            Error::BrokenTransportSend(ref se) => f.write_str("The underlying transport failed while attempting to send a response"),
            Error::Service(ref se) => f.write_str("The underlying service failed to process a request"),
```
Or we have them behave like transparent errors (using the `thiserror` crate vocabulary here), in which case we should forward both source and Display/description.

If we don't then walking the error chain (through `source()`) displays twice the same description.

File: src/error.rs [hunk 1]
```diff
@@ -66,28 +68,16 @@ where
 impl<T, I> error::Error for Error<T, I>
 where
     T: Sink<I> + TryStream,
-    <T as Sink<I>>::Error: error::Error,
-    <T as TryStream>::Error: error::Error,
+    <T as Sink<I>>::Error: error::Error + 'static,
```
Comment (jonhoo) [inline src/error.rs:71]: Ah, this is a backwards-incompatible change, which means we'll need to bump the major version. Just an OBS to whomever ends up pushing this change out.

**Review by jonhoo — APPROVED** [review]
Yes, this looks right to me! Also matches with the new [guidance from the Error Working Group](https://blog.rust-lang.org/inside-rust/2021/07/01/What-the-error-handling-project-group-is-working-towards.html#guidelines-for-implementing-displayfmt-and-errorsource).

**Review by LucioFranco — APPROVED** [review]
I won't have time to push this out this week but this change LGTM

**Comment by jonhoo** [comment]
Published as 0.6.0 :tada:

**Pull request merged**